### PR TITLE
Newrelic

### PR DESCRIPTION
**Pull request by mbogor**
#### What's this PR do?
setting up newrelic on cronjobs
#### Where should the reviewer start?
jobs directory
#### How should this be manually tested?
start the app and check newrelic
#### What are the relevant tickets?
AT-18

### CODE REVIEW

File: jobs/scheduling/TimeOffRequests.js [hunk 1]
```diff
@@ -42,8 +44,10 @@ function filterRequestsAndHandleShifts (allRequests) {
   });
 
   //For each pending request, look for any shifts that fall within the time off request
-  newRequests.forEach(retrieveOverlappingShifts);
-
+  newRequests.forEach((request, index)=>{
+    retrieveOverlappingShifts(request);
+    if(index===newRequests.length-1) nr.endTransaction();
+  })
```
Comment (apackin) [inline jobs/scheduling/TimeOffRequests.js]: bluebird.map(newRequests, retrieveOverlappingShifts).then(nr.endTransaction);

File: cache.js [hunk 2]
```diff
@@ -21,6 +21,7 @@ function WiWUsersCache(){
           else if ((/canonicalEmail/.test(user.notes) && user.is_deleted)) hash.deleted[JSON.parse(user.notes).canonicalEmail] = user; 
           else if (user.is_deleted) hash.deleted[user.email] = user;
         });
+        CONSOLE_WITH_TIME(`Active users: ${Object.keys(hash).length}`);
```
Comment (apackin) [inline cache.js]: length - 1 } for the .deleted key ?

File: config.js [hunk 3]
```diff
@@ -1,7 +1,6 @@
 'use strict';
 const moment = require('moment-timezone');
 moment.tz.setDefault("America/New_York");
-
```
Comment (apackin) [inline config.js]: Do we need this in our commit history?

File: jobs/recruitment/volunteerMatches.js [hunk 4]
```diff
@@ -1,13 +1,20 @@
...
+const nr = require('newrelic');
 const CronJob = require('cron').CronJob;
 const request = require('request-promise');
 const ctlOnline = require('../../ctlOnline.js');
 const mailChimp = require('./mailChimp.js');
 
-new CronJob(CONFIG.time_interval.request_VolunteerMatches, reconcileVMRecs, null, true);
+new CronJob(CONFIG.time_interval.request_VolunteerMatches, newRelicMonit, null, true);
+
+function newRelicMonit(){
```
Comment (apackin) [inline jobs/recruitment/volunteerMatches.js]: 💯

File: jobs/recruitment/volunteerMatches.js [hunk 5]
```diff
@@ -1,13 +1,20 @@
...
+  const monitoringEnd = function(){reconcileVMRecs().then(()=>{nr.endTransaction()})};
+  nr.createBackgroundTransaction('recruitment: volunteerMatches', monitoringEnd)();
```
Comment (apackin) [inline jobs/recruitment/volunteerMatches.js]: Should this happen first?

File: jobs/recruitment/volunteerMatches.js [hunk 6]
```diff
@@ -1,13 +1,20 @@
...
 const CronJob = require('cron').CronJob;
 const request = require('request-promise');
 const ctlOnline = require('../../ctlOnline.js');
 const mailChimp = require('./mailChimp.js');
 
-new CronJob(CONFIG.time_interval.request_VolunteerMatches, reconcileVMRecs, null, true);
+new CronJob(CONFIG.time_interval.request_VolunteerMatches, newRelicMonit, null, true);
+
+function newRelicMonit(){
+  const monitoringEnd = function(){reconcileVMRecs().then(()=>{nr.endTransaction()})};
```
Comment (apackin) [inline jobs/recruitment/volunteerMatches.js]: .then(()=>{nr.endTransaction()} === .then(nr.endTransaction}

File: jobs/recruitment/volunteerMatches.js [hunk 7]
```diff
@@ -22,6 +29,11 @@ function subNewVMToMailchimp () {
     return Promise.all([
                        ...mailChimp.postCSV(CSV)
                        ]);
+  })
+  // .then(returnArr => {
```
Comment (apackin) [inline jobs/recruitment/volunteerMatches.js]: Adding these comments lines?

I think we should start adding ${err.stack} to our catch blocks so we can see the stack trace for error that triggered the catch. 
(Regarding our past convos on this: unfortunately doing a console.trace within the catch block traces the stack of getting the catch which is not the trace of the error that triggered the catch :-/ )

File: jobs/scheduling/AttendanceSync.js [hunk 8]
```diff
@@ -1,17 +1,16 @@
 'use strict';
-var CronJob = require('cron').CronJob;
-var canvas = require('../../canvas.js');
-var bignumJSON = require('json-bignum');
-var Request = require('request');
-var mandrill = require('mandrill-api/mandrill');
-var mandrill_client = new mandrill.Mandrill(KEYS.mandrill.api_key);
-var fs = require('fs');
+const CronJob = require('cron').CronJob;
```
Comment (apackin) [inline jobs/scheduling/AttendanceSync.js]: ES6 for eva! (But really just until ES7 come out)

File: jobs/scheduling/AttendanceSync.js [hunk 9]
```diff
@@ -119,10 +118,11 @@ function findCanvasUsersByEmail(GTWusers){
         return;}
       var userID = canvasUsers[0].id;
       queryForCanvasCoursesAndAssignments(userID);
+
```
Comment (apackin) [inline jobs/scheduling/AttendanceSync.js]: empty line ^^

File: jobs/scheduling/AttendanceSync.js [hunk 10]
```diff
@@ -119,10 +118,11 @@ function findCanvasUsersByEmail(GTWusers){
...
     })
     .catch(function(){
       CONSOLE_WITH_TIME('Call to Canvas for user', GTWUser.email, 'has failed.')
-    })
+    });
```
Comment (apackin) [inline jobs/scheduling/AttendanceSync.js]: :1st_place_medal: ;

File: jobs/scheduling/MergeOpenShifts.js [hunk 11]
```diff
@@ -59,7 +61,10 @@ function processDataAndMakeMergeAPICalls(data) {
     WhenIWork.post('batch', batchPayload, function(response) {
       //The response at this point doesn't include error status codes so we're looking for a message that indicates an error
       if (response && response.message && /error/.test(response.message)) CONSOLE_WITH_TIME("[ERROR 61] in Job MergeOpenShifts Batch Response: ", response);
-      else (CONSOLE_WITH_TIME("Job MergeOpenShifts Batch Post Success"));
+      else {
+        CONSOLE_WITH_TIME("Job MergeOpenShifts Batch Post Success");
+        nr.endTransaction();
```
Comment (apackin) [inline jobs/scheduling/MergeOpenShifts.js]: nice placement

File: jobs/scheduling/NotifyMoreShifts.js [hunk 12]
```diff
@@ -40,6 +44,7 @@ function notifyMoreShifts() {
         .then(function(shiftsToSup){
           mandrillEachUser(shiftNotificationResult.usersBeingNotified, shiftsToSup);
         })
+        .then(()=>{nr.endTransaction();})
```
Comment (apackin) [inline jobs/scheduling/NotifyMoreShifts.js]: Move this into newRelicMonit like in the VM script

File: runJobs.js [hunk 13]
```diff
@@ -3,23 +3,29 @@ global.KEYS = require('./keys.js');
...
 }
 
 function CacheInit(){
-  cache()
+  return cache()
   .then(function(users){
     global.USERS_CACHE = users;
     CONSOLE_WITH_TIME('Cache Initialized or Refreshed in runJobs');
-    const jobs = require('require-all')(__dirname + '/jobs');
+    const jobs = require('require-all')(__dirname + '/jobs/training');
```
Comment (apackin) [inline runJobs.js]: 😞

**Comment by apackin** [comment]
Well done. Excited to see this merged in. Made some comments about minor errors and testing code slipping through - major changes look :1st_place_medal:

File: cache.js (hunk 2 above)
Comment (mbogor) [inline cache.js]: true

File: jobs/recruitment/volunteerMatches.js (hunk 6 above)
Comment (mbogor) [inline jobs/recruitment/volunteerMatches.js]: yeah, i wrote it like that to be safer, i think i'll keep it this way to be sure

File: jobs/recruitment/volunteerMatches.js (hunk 7 above)
Comment (mbogor) [inline jobs/recruitment/volunteerMatches.js]: you're right on the err.stack, i don't want to add it to this pr though. and these lines were commented out before i touched the script :/

File: jobs/scheduling/AttendanceSync.js (hunk 8 above)
Comment (mbogor) [inline jobs/scheduling/AttendanceSync.js]: lol

**Pull request merged**